src/expressions.js

```diff
@@ -40,3 +40,42 @@ exports.satisfiesExpression = function(scopeset, expr) {
     exports.satisfiesExpression(scopeset, subexpr)
   );
 };
+
+/**
+ * Given a scopeset and a scope-expression that it
+ * does not satisfy, explain why the scope-expression was
+ * not satisfied.
+ *
+ * This is separate from satisfiesExpression in order to keep
+ * that code path simple and clean due to its important nature.
+ *
+ * Returns a scope expression where all scopes that exist are
+ * missing from the scopeset. Any scopes under an AllOf key
+ * are definitely needed to satisfy the expression and at least
+ * one of the scopes under an AnyOf must be provided to satisfy.
+ */
+exports.explainFailure = function(scopeset, expr) {
+  assert(Array.isArray(scopeset), 'Scopeset must be an array.');
+  assert(exports.validExpression(expr));
+
+  if (exports.satisfiesExpression(scopeset, expr)) {
+    return null;
+  }
+
+  const operator = Object.keys(expr)[0];
+  const subexpressions = expr[operator];
+
+  return {[operator]: subexpressions.reduce((explain, subexpr) => {
+    if (typeof subexpr === 'string') {
+      if (operator === 'AnyOf' || !scopeset.some(scope => patternMatch(scope, subexpr))) {
+        explain.push(subexpr);
+      }
+    } else if (operator === 'AnyOf' || !exports.satisfiesExpression(scopeset, subexpr)) {
+      const subexplain = exports.explainFailure(scopeset, subexpr);
+      if (subexplain) {
+        explain.push(subexplain);
+      }
+    }
+    return explain;
+  }, [])};
```

Comment:
Maybe consider using map/filter/some...

```js
// reduceExpression will reduce expr as much as possible assuming scopes given are already satisfied.
// returns null, if expr is satisfied.
const reduceExpression = (expr, scopes) => {
  if (typeof expr === 'string') {
    if (scopes.some(s => patternMatch(scope, expr))) {
       return null;
    }
    return expr;
  }

  if (expr.AllOf) {
    const AllOf = expr.AllOf.map(e => reduceExpression(e, scopes)).filter(e => e !== null);
    if (AllOf.length === 0) {
      return null;
    }
    return {AllOf};
  }

  const AnyOf = expr.AnyOf.map(e => reduceExpression(e, scopes));
  if (AnyOf.some(e => e === null)) {
    return null;
  }
  return {AnyOf};
};
```

Reply:
How strongly do you feel about using your way instead of mine? I kinda like mine a tiny bit more but am happy to switch if you feel strongly

Reply:
Not much... in fact we could probably do something even simpler...

I do somewhat care that we try to use `map/filter/some/every` on arrays... and that we treat each of the 3 cases individually, as this makes it easier to review correctness.

Reply:
@imbstack, I know that you only see this function as being used to provide explanations.

But this could be used to reduce a scope-expression prior to sending it off to tc-auth...
I don't immediately have a use-case, but I can totally imagine this being used in security sensitive scenarios, where you want to partially factor out some requirements from a scope-expression...

Reply:
I just feel like the `satisfiesExpression` function is so nice and simple. Adding this mess into any secure part of the system seems like a footgun that I can't think of any _great_ reason to do. Definitely we can speculate about options, but I would think we want to avoid doing that until we have a great reason.

Reply:
fair point :)